dask/dataframe/io/parquet/fastparquet.py

```diff
@@ -288,154 +278,408 @@ def read_metadata(
-
-        # Loop over DataFrame partitions.
-        # Each `part` will be a row-group or path ()
-        for i, part in enumerate(partsin):
-            if pqpartitions and fast_metadata:
-                # Case (3A)
-                # We can pass a "_metadata" path
-                file_path = base_path + "_metadata"
-                i_path = i
-            elif (
-                pf_deps
-                and isinstance(part.columns[0].file_path, str)
-                and not pqpartitions
+        return meta, dtypes, index_cols, categories_dict, categories, index
+
+    @classmethod
+    def _update_metadata_options(
+        cls,
+        pf,
+        parts,
+        gather_statistics,
+        split_row_groups,
+        index_cols,
+        filters,
+    ):
+        # Cannot gather_statistics if our `parts` is alreadey a list
+        # of paths, or if we are building a multiindex (for now).
+        # We also don't "need" to gather statistics if we don't
+        # want to apply any filters or calculate divisions.
+        if (
+            isinstance(parts, list) and len(parts) and isinstance(parts[0], str)
+        ) or len(index_cols) > 1:
+            gather_statistics = False
+        elif filters is None and len(index_cols) == 0:
+            gather_statistics = False
+
+        # Make sure gather_statistics allows filtering
+        # (if filters are desired)
+        if filters:
+            # Filters may require us to gather statistics
+            if gather_statistics is False and pf.info.get("partitions", None):
+                warnings.warn(
+                    "Filtering with gather_statistics=False. "
+                    "Only partition columns will be filtered correctly."
+                )
+            elif gather_statistics is False:
+                raise ValueError("Cannot apply filters with gather_statistics=False")
+            elif not gather_statistics:
+                gather_statistics = True
+
+        # Determine which columns need statistics.
+        flat_filters = _flatten_filters(filters)
+        stat_col_indices = {}
+        for i, name in enumerate(pf.columns):
+            if name in index_cols or name in flat_filters:
+                stat_col_indices[name] = i
+
+        # If the user has not specified `gather_statistics`,
+        # we will only do so if there are specific columns in
+        # need of statistics.
+        if gather_statistics is None:
+            gather_statistics = len(stat_col_indices.keys()) > 0
+        if split_row_groups is None:
+            split_row_groups = False
+
+        return (
+            gather_statistics,
+            split_row_groups,
+            stat_col_indices,
+        )
+
+    @classmethod
+    def _organize_row_groups(
+        cls,
+        pf,
+        split_row_groups,
+        gather_statistics,
+        stat_col_indices,
+        filters,
+        dtypes,
+        base_path,
+        paths,
+    ):
+        """Organize row-groups by file."""
```

Comment:
(fastparquet will *write* one row-group per file)

Reply:
cudf writes files with row-groups of ~128MB because a single GPU can be used to read multiple row-groups concurrently.  PyArrow also allows the user to specify the row-group size at write time.  Therefore, we see (and highly recommend) datasets with multiple row-groups per file.

Reply:
But why multiple row groups in the same file?
An issue to do this could be created for fastparquet, with low priority.

Reply:
For the cudf/dask_cudf case - It is most optimal for each output/partition to contain multiple row-groups (so that the GPU is saturated).  Right now dask does not allow multiple files to be read in as a single output partition, so the only way to do this is to have multi-row-group files.

Also, an optimal partition size in `dask_cudf` depends on the device memory.  Therefore, we generate datasets with many-row-group files to allow the same dataset to be processed efficiently on different hardware.